### ISSUE TITLE
Some sort of QAT calibration pipeline to make smaller tensor formats viable and push models under 4 bpw

### ISSUE THREAD

**Issue by Not-a-web-Developer** [notactuallytreyanastasio/shoehorn#1]
tl;dr: maybe we can make bonsai at home, because why not

I've been using the [binary bonsai 27B](https://docs.prismml.com/models/bonsai-27b) on my 3060 for a while now, it's not the fastest but it's shockingly good for what it is; that said, prismML don't really talk about exactly what they did to train qwen 3.6 27B to outperform a regular IQ2_XXS quant at 1.125 bpw (for the binary version of bonsai, which is the one i run), but i'm assuming it's some sort of QAT done in a loop to train a student model with conveniently aligned weights that can relatively easily be crunched down to something as small as Q1_0. 

It's not the most resource-efficient thing to do, but I think there's optimisations that can be made to fit into shoehorn better: 

1. the corpus generated by the original model weights to build the imatrix tensor can also be used to do the QAT and distillation provided the corpus is large enough;
2. the teacher model doesn't need to stay resident in VRAM or RAM, if you can figure out a way to first do generation of everything needed from the teacher model (logits, other stuff I probably know nothing about because I'm not from this field) and then use it instead of having to have a teacher to generate everything at the same time as it's needed;
3. the student model doesn't need to be entirely in VRAM either, pretty sure there's ways to do much slower training per-layer, though that should either be a flag or part of the estimation checks shoehorn already does;
4. the vision tower should probably not be pushed down below 4 bpw, and in general anything that's not being pushed below 4 bpw should be frozen and kept out of QAT and distillation

this is just a rough mishmash of ideas i thought of after far too much reading of things my agent recommended me to read which were far beyond my ability to understand (and obviously ESL warning), so my apologies if this sounds like jargon soup; i don't think this is very good first issue, but i hope it's coherent enough to meaningfully help your agent?
